keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+current list is not exhaustive, and we do not endeavor to develop an exhaustive list prior to releasing the 
+Application Kind. We expect users to use the fields of ObjectMeta (i.e. Labels and Annotations) for custom use cases. 
+When the community forms a consensus that a particular label or annotation has become prevalent across a large 
+number of tools, we will promote that metadata to a field. Note that this can be done in a backward compatible manner 
+without respect to the stability of the Application Kind.
+
+#### Validation
+Upon Application creation, The API Server shall perform the following validation.
+
+1. The API Server shall validate that the Application is a valid ObjectMeta with a name that is a valid DNS subdomain. 
+In effect, the API Server will simply validate that the Application is a valid Kubernetes resource.
+1. The API Server will validate that the Application's `Spec.Selector` is a valid, non-nil, set-based Selector.
+1. If the Application has a `.Spec.HealthCheck`, the API Server will validate that the `Template` of the 
+ApplicationHealthCheck is a valid Pod with a readiness check.
+
+The API Server shall perform the following validation upon Application mutation.
+
+1. The API server shall ensure that the new version of the Application is valid using the above definition of a valid 
+newly created Application instance. 
+1. The API server shall ensure that the Applications `Spec.Selector` has not been mutated. This latter condition has 
+been implemented in many places in the Kubernetes API to prevent unintentional orphaning. This is our purpose here.
+
+#### Generation Incrementation
+
+1. Upon creation of an Application object, the API Server shall set the initial value of the `Generation` field to 0.
+1. The API Server shall, upon update of an Application object, increment the `Generation` of the Application object by 1. 
+
+
+#### Namespaces
+All components and dependencies of an Application must reside in the same namespace as the Application. Given the 
+design of Kubernetes namespaces, we feel this is a reasonable, and desirable, constraint. A user, who has write 
+permissions for a particular namespace, can install an Application and all of its components into the same namespace, 
+and only those users with the necessary permissions can modify the Application. Additionally, administrators may limit 
+the resources consumed by an Application by applying quotas to the namespace in which it resides.
+
+#### Composition (Is A)
+The Application Kind is designed to handle the composition of Kubernetes primitive Kinds. This includes CRDs 
+(Custom Resource Definitions). The manifest below demonstrates how this can be applied to the WordPress CMS.
+
+
+```yaml
+apiVersion: v1
+Kind: Application
+metadata:
+  name: wordpress
+spec:
+  type: wordpress
+  version: 1.0
+  url: 
+  selector:
+    matchLables:
+      app: wordpress
```

Comment:
correct me if i'm wrong, but w/ this api, the application component is going to select all services w/ a label named "wordpress" (and all statefulsets).

Or the selected resources are additionally filtered by the component names in the components map?  (I would expect the latter but I haven't seen it clearly spelled out)

Reply:
actually it looks like this spec also introduces a "comonent" label, so i guess it filters on:
resource=resourcetype
app=appname
component=componentname

to construct the set of app resources.  Again would be good to clearly state somewhere how this app resource selection is intended to work.

Reply:
ok no, the component label doesn't appear to be used by this spec.

Which makes me wonder about the purpose of the "components" datastructure.  All that seems to actually be used is a label selector + a list of resource groupversionkinds.  The actual resource names are not used by any part of the spec that I saw, so i don't know what purpose they serve in the data structure.

Reply:
I added a better explanation in the CRD type definition. Getting things by name might work if we are okay with every instance of an application mapping distinct names to the GV of the object.